### DISCUSSION TITLE
Cursors look blurry

### DISCUSSION THREAD

**Discussion by MsBlockCat** [opentoonz/opentoonz#6752] [General]
### Issue Summary
This may be a macOS specific issue, or it may be because I have a 2560 x 1600 display, I haven't done enough testing to figure that out.
On my MacBook Pro, OpenToonz's cursors all appear blurry on screen.

### Expected Results
All the cursors look nice and HD

### Actual Results
They all look blurry

### System Information
* **OpenToonz Version**: 1.4.0
* **Operating System**: macOS Catalina
* **CPU**: Quad core i5
* **Memory**: 8 GB LPDDR3
* **Graphics Card**: 
* **Graphics Tablet**:

**Comment by MsBlockCat** [comment]
Upon testing with some different displays on Windows, this seems to also happen on Windows. I think it's only a problem with higher-resolution displays, though

**Comment by MsBlockCat** [comment]
Upon further investigation, it seems the issue is simply that the cursor icons (found [here](https://github.com/opentoonz/opentoonz/tree/master/toonz/sources/tnztools/Resources)) are too low resolution. If higher resolution versions were put in would it be possible to make them appear the same size on screen, rather than bigger which I assume is what would happen by default? Or would higher resolution cursors scale with the current code so they'd be the same size as the current ones?

**Comment by MsBlockCat** [comment]
If it's possible to make the code work with higher resolution cursors, I'd be willing to help make the cursors more HD! As a side effect I'd fix the red dot on the bottom left of the pan.png cursor icon that I just noticed, I'm assuming that's unintentional right?

**Comment by MsBlockCat** [comment]
<img width="32" alt="pan" src="https://user-images.githubusercontent.com/37186692/76153315-478deb00-6098-11ea-8149-db9d5d44b91d.png">

Here's an example of a more HD pan.png cursor icon, if we decide to go that route. I pasted it on top of a screenshot of OpenToonz, and on my high res display it looks way nicer! (it's the right one there)
![Screen Shot 2020-03-07 at 5 11 58 PM](https://user-images.githubusercontent.com/37186692/76153328-760bc600-6098-11ea-98ec-ddc4deee7bd1.png)

**Comment by MsBlockCat** [comment]
I'm also doubling the DPI along with the length and height so the icon should stay the same size, but it's just a question of if OpenToonz handles that well in the code which I'm not sure if it does or not

**Comment by MsBlockCat** [comment]
Here's what I've got so far, along with notes of tweaks I made or other things that might come in handy. I'll wait to make more until we're sure we can actually use these in OpenToonz, these take a while to upscale!

Notes:
- SD Normal cursor tip coordinates: 11 x 25
- HD Normal cursor tip coordinates: 22 x 51

Fixed:
- Removed red dot from bottom left of pan.png
- Made brush_crosshair.png centered
- brush_crosshair.png, brush_large.png, edit_FX_notext.png, ex_area_left.png, ex_freehand_left.png, ex_hv.png, and ex_hv_left.png went from 96 dpi to 144 dpi rather than the normal 72 dpi to 144 dpi (should this be changed to 192 dpi?)
- ex_FX.png, ex_FX_left.png, and ex_fill_no_autopaint_left.png went from 120 DPI to 144 DPI (should this be changed to 240 DPI?)
- ex_area_left.png, ex_freehand_left.png, and ex_hv_left.png shifted right slightly to be symmetrical with right
- Added white border to ex_area.png, ex_area_left.png, and ex_freehand.png
- Made F in ex_freehand.png and ex_freehand_left.png shorter
- Made center of crosshairs transparent if they weren’t

Concerns:
- All the ex_SomeName.jpg files were 35 x 30 pixels, except for ex_FX.jpg and ex_FX_left.jpg
- All the images besides ex_SomeName.jpg ones were 32 x 32, besides fill.png which was 35 x 30

<img width="32" alt="bender" src="https://user-images.githubusercontent.com/37186692/76165993-a6ec0980-6131-11ea-8263-a27a12c7f9b8.png">
<img width="32" alt="brush_crosshair" src="https://user-images.githubusercontent.com/37186692/76165994-a6ec0980-6131-11ea-8030-0df27193bd6a.png">
<img width="32" alt="brush_large" src="https://user-images.githubusercontent.com/37186692/76165995-a6ec0980-6131-11ea-94e6-60c78f7be40a.png">
<img width="32" alt="brush" src="https://user-images.githubusercontent.com/37186692/76165996-a6ec0980-6131-11ea-820b-42389d720d29.png">
<img width="32" alt="cutter" src="https://user-images.githubusercontent.com/37186692/76165997-a784a000-6131-11ea-932f-3e0e00d09f38.png">
<img width="32" alt="disable" src="https://user-images.githubusercontent.com/37186692/76166109-604adf00-6132-11ea-99bb-94c2b769bda2.png">
<img width="32" alt="edit_FX_notext" src="https://user-images.githubusercontent.com/37186692/76165999-a784a000-6131-11ea-9254-39e6f5197d5a.png">
<img width="32" alt="edit_FX" src="https://user-images.githubusercontent.com/37186692/76166000-a784a000-6131-11ea-8bd7-5dae53c182ae.png">
<img width="32" alt="eraser" src="https://user-images.githubusercontent.com/37186692/76166001-a784a000-6131-11ea-973b-a2b2b8f7298a.png">
<img width="35" alt="ex_area_left" src="https://user-images.githubusercontent.com/37186692/76166002-a81d3680-6131-11ea-9a93-faefc5b17f3a.png">
<img width="35" alt="ex_area" src="https://user-images.githubusercontent.com/37186692/76166003-a81d3680-6131-11ea-8786-4988f97cf657.png">
<img width="35" alt="ex_fill_no_autopaint_left" src="https://user-images.githubusercontent.com/37186692/76166005-a81d3680-6131-11ea-877c-f10131df4f1c.png">
<img width="35" alt="ex_fill_no_autopaint" src="https://user-images.githubusercontent.com/37186692/76166006-a81d3680-6131-11ea-9078-4b3520924df5.png">
<img width="35" alt="ex_freehand_left" src="https://user-images.githubusercontent.com/37186692/76166007-a81d3680-6131-11ea-8d62-182d67b1865b.png">
<img width="35" alt="ex_freehand" src="https://user-images.githubusercontent.com/37186692/76166008-a8b5cd00-6131-11ea-9129-90b4eb7486fe.png">
<img width="32" alt="ex_FX" src="https://user-images.githubusercontent.com/37186692/76166009-a8b5cd00-6131-11ea-9761-bd3ea296cde3.png">
<img width="32" alt="ex_FX_left" src="https://user-images.githubusercontent.com/37186692/76166010-a8b5cd00-6131-11ea-85ed-39712859d05c.png">
<img width="32" alt="ex_hv_left" src="https://user-images.githubusercontent.com/37186692/76166011-a8b5cd00-6131-11ea-9c2c-5831663b49cb.png">
<img width="32" alt="ex_hv" src="https://user-images.githubusercontent.com/37186692/76166012-a8b5cd00-6131-11ea-8d26-988845028600.png">
<img width="35" alt="fill" src="https://user-images.githubusercontent.com/37186692/76166013-a94e6380-6131-11ea-9e02-4df499a14fa9.png">
<img width="32" alt="pan" src="https://user-images.githubusercontent.com/37186692/76166014-a94e6380-6131-11ea-9877-0952778f6319.png">
<img width="32" alt="stroke_select" src="https://user-images.githubusercontent.com/37186692/76166015-a94e6380-6131-11ea-8753-41ebdfb1d83c.png">
<img width="32" alt="zoom" src="https://user-images.githubusercontent.com/37186692/76166016-a94e6380-6131-11ea-946d-e6bcd510292c.png">

**Comment by RodneyBaker** [comment]
You've been busy!

A question for the experts:  Can cursor icons be SVG?
If yes, then they could be resolution independent by being SVG.

It is my understanding that most regular icons in Opentoonz have been transitioned to SVG.
Perhaps the icons you've noted as blurry in other report are ones that have not yet been converted?

**Comment by MsBlockCat** [comment]
That was my thought, that the blurry icons are likely not SVG, but I haven't been able to find the source of those despite looking around in the code and online, so I'm not 100% sure.

As for making the cursors SVG, I'm not an expert so I'll leave that question to those who are haha

**Comment by MsBlockCat** [comment]
Oh yeah, this probably doesn't need the macOS tag as this is a cross-platform issue, and just is only visible on high resolution displays

**Comment by artisteacher** [comment]
You can find the icons in toonz/sources/toonz/Resources

**Comment by MsBlockCat** [comment]
Interesting... What does toonz/sources/tnztools/Resources do then?

**Comment by MsBlockCat** [comment]
I'm willing to keep working on this, I'll just need to know if they should be .svg or .png first!

**Comment by MsBlockCat** [comment]
If I'm interpreting [toonz/sources/tnztools/cursormanager.cpp line 152 and that area](https://github.com/opentoonz/opentoonz/blob/5d6d967c5fdfb975b182e440fb8c4a91548f5919/toonz/sources/tnztools/cursormanager.cpp#L142-L162) correctly, it looks like unfortunately only .png would work, unless that bit was rewritten, at least. It probably is possible to rewrite those few lines, but I have no idea how!

**Comment by RodneyBaker** [comment]
It looks to me like it is relying on the alpha channel of PNG files.
So I think you have your answer.  It needs to be PNG.

**Comment by MsBlockCat** [comment]
I tested just putting in the new icons I have so far and building without editing any code, and they show up way too big (twice as big in x and twice as big in y, so 4x as big total)! I was worried that might happen, but this leads me to believe there are two parts to this problem: Making the new icons, which I can do, but also the second part which is editing the code to make the cursors show up the right size, which should be halving the x and halving the y of it. [This](https://github.com/opentoonz/opentoonz/blob/5d6d967c5fdfb975b182e440fb8c4a91548f5919/toonz/sources/tnztools/cursormanager.cpp) most likely holds the key to that second step, but I'm not sure how exactly. I'll keep looking into the code aspect, but I don't really know C++ so I likely won't get far.

Edit: [This](https://doc.qt.io/archives/qt-4.8/qpixmap.html#scaled) and [this](https://doc.qt.io/archives/qtjambi-4.5.2_01/com/trolltech/qt/gui/QPixmap.html#scaled(com.trolltech.qt.core.QSize,%20com.trolltech.qt.core.Qt.AspectRatioMode)) from Qt's documentation look potentially useful.
